sync/blocksync/syncer_test.go

```diff
@@ -0,0 +1,319 @@
+	client  *syncclient.TestClient
+	blocks  []*types.Block
+}
+
+// newTestEnvironment creates a new test environment with generated blocks
+func newTestEnvironment(t *testing.T, numBlocks int) *testEnvironment {
+	t.Helper()
+
+	var (
+		key1, _        = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+		key2, _        = crypto.HexToECDSA("8a1f9a8f95be41cd7ccb6168179afb4504aefe388d1e14474d32c45c72ce7b7a")
+		addr1          = crypto.PubkeyToAddress(key1.PublicKey)
+		addr2          = crypto.PubkeyToAddress(key2.PublicKey)
+		genesisBalance = big.NewInt(1000000000)
+		signer         = types.HomesteadSigner{}
+	)
+
+	// Ensure that key1 has some funds in the genesis block.
+	gspec := &core.Genesis{
+		Config: &params.ChainConfig{HomesteadBlock: new(big.Int)},
+		Alloc:  types.GenesisAlloc{addr1: {Balance: genesisBalance}},
+	}
+	engine := dummy.NewETHFaker()
+
+	_, blocks, _, err := core.GenerateChainWithGenesis(gspec, engine, numBlocks, 0, func(i int, gen *core.BlockGen) {
+		// Generate a transaction to create a unique block
+		tx, _ := types.SignTx(types.NewTransaction(gen.TxNonce(addr1), addr2, big.NewInt(10), params.TxGas, nil, nil), signer, key1)
+		gen.AddTx(tx)
+	})
+	require.NoError(t, err)
+
+	// The genesis block is not include in the blocks slice, so we need to prepend it
+	blocks = append([]*types.Block{gspec.ToBlock()}, blocks...)
+
+	return &testEnvironment{
+		chainDB: rawdb.NewMemoryDatabase(),
+		blocks:  blocks,
+	}
+}
+
+// withCustomBlockProvider sets up the test client with a custom block provider function
+func (e *testEnvironment) withCustomBlockProvider(onGetBlock func(hash common.Hash, height uint64) *types.Block) *testEnvironment {
+	blockProvider := &handlers.TestBlockProvider{GetBlockFn: onGetBlock}
+	blockHandler := handlers.NewBlockRequestHandler(
+		blockProvider,
+		message.Codec,
+		handlerstats.NewNoopHandlerStats(),
+	)
+	e.client = syncclient.NewTestClient(
+		message.Codec,
+		nil,
+		nil,
+		blockHandler,
+	)
+	return e
+}
+
+// withDefaultBlockProvider sets up the test client with a provider that returns all blocks
+func (e *testEnvironment) withDefaultBlockProvider() *testEnvironment {
+	return e.withCustomBlockProvider(func(hash common.Hash, height uint64) *types.Block {
+		if height > uint64(len(e.blocks)) {
+			return nil
+		}
+		block := e.blocks[height]
+		if block.Hash() != hash {
+			return nil
+		}
+		return block
+	})
+}
+
+// prePopulateBlocks writes some blocks to the database before syncing (by block height)
+func (e *testEnvironment) prePopulateBlocks(blockHeights ...int) *testEnvironment {
+	batch := e.chainDB.NewBatch()
+	for _, height := range blockHeights {
+		if height <= len(e.blocks) {
+			// blocks[0] is block number 1, blocks[1] is block number 2, etc.
+			block := e.blocks[height]
+			rawdb.WriteBlock(batch, block)
+			rawdb.WriteCanonicalHash(batch, block.Hash(), block.NumberU64())
+		}
+	}
+	require.NoError(nil, batch.Write())
```

Comment:
We should not be passing `nil` in to any of these functions.

We could (in preferred order)
- return the error to the caller
- pass in the testing struct
- explicitly panic

Having a nil pointer in the require library isn't ideal.